File: src/entities/item/itemProperties.ts
const ITEM_PROPERTY_CONFIG = {
  ammunition: {
    labels: { en: "Ammunition", pt: "Municao" },
    aliases: ["ammunition", "ammo", "municao"],
  },
  finesse: {
    labels: { en: "Finesse", pt: "Acuidade" },
    aliases: ["finesse", "acuidade"],
  },
  heavy: {
    labels: { en: "Heavy", pt: "Pesada" },
    aliases: ["heavy", "pesada", "pesado"],
  },
  light: {
    labels: { en: "Light", pt: "Leve" },
    aliases: ["light", "leve"],
  },
  loading: {
    labels: { en: "Loading", pt: "Carregamento" },
    aliases: ["loading", "carregamento", "recarga"],
  },
  range: {
    labels: { en: "Range", pt: "Alcance" },
    aliases: ["range", "alcance", "ranged", "distancia", "a distancia"],
  },
  reach: {
    labels: { en: "Reach", pt: "Alcance estendido" },
    aliases: ["reach", "alcance estendido"],
  },
  special: {
    labels: { en: "Special", pt: "Especial" },
    aliases: ["special", "especial"],
  },
  thrown: {
    labels: { en: "Thrown", pt: "Arremesso" },
    aliases: ["thrown", "arremesso", "arremessavel"],
  },
  two_handed: {
    labels: { en: "Two-Handed", pt: "Duas maos" },
    aliases: ["two_handed", "two handed", "two-handed", "duas maos"],
  },
  versatile: {
    labels: { en: "Versatile", pt: "Versatil" },
    aliases: ["versatile", "versatil"],
  },
  stealth_disadvantage: {
    labels: { en: "Stealth disadvantage", pt: "Desvantagem em furtividade" },
    aliases: [
      "stealth_disadvantage",
      "stealth disadvantage",
      "disadvantage on stealth",
      "desvantagem em furtividade",
    ],
  },
} as const;

export type ItemPropertySlug = keyof typeof ITEM_PROPERTY_CONFIG;

export const ITEM_PROPERTY_SLUGS = Object.keys(
  ITEM_PROPERTY_CONFIG,
) as ItemPropertySlug[];

const normalizePropertyToken = (value: string) =>
  value
    .trim()
    .toLowerCase()
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/[_-]+/g, " ")
    .replace(/\s+/g, " ");

const ITEM_PROPERTY_ALIAS_TO_SLUG = Object.entries(ITEM_PROPERTY_CONFIG).reduce(
  (accumulator, [slug, config]) => {
    config.aliases.forEach((alias) => {
      accumulator[normalizePropertyToken(alias)] = slug as ItemPropertySlug;
    });
    accumulator[normalizePropertyToken(slug)] = slug as ItemPropertySlug;
    return accumulator;
  },
  {} as Record<string, ItemPropertySlug>,
);

type NormalizedPropertiesResult =
  | { ok: true; value: ItemPropertySlug[]; invalid: [] }
  | { ok: false; value: ItemPropertySlug[]; invalid: string[] };

export const resolveItemPropertySlug = (value: string): ItemPropertySlug | null => {
  if (!value.trim()) {
    return null;
  }
  return ITEM_PROPERTY_ALIAS_TO_SLUG[normalizePropertyToken(value)] ?? null;
};

export const normalizeItemProperties = (
  values: readonly string[] | null | undefined,
): NormalizedPropertiesResult => {
  const normalized: ItemPropertySlug[] = [];
  const invalid: string[] = [];
  const seen = new Set<ItemPropertySlug>();

  for (const entry of values ?? []) {
    const rawValue = String(entry ?? "").trim();
    if (!rawValue) {
      continue;
    }

    const slug = resolveItemPropertySlug(rawValue);
    if (!slug) {
      if (!invalid.includes(rawValue)) {
        invalid.push(rawValue);
      }
      continue;
    }

    if (!seen.has(slug)) {
      seen.add(slug);
      normalized.push(slug);
    }
  }

  if (invalid.length > 0) {
    return { ok: false, value: normalized, invalid };
  }

  return { ok: true, value: normalized, invalid: [] };
};

export const parseItemPropertiesInput = (
  input: string,
): NormalizedPropertiesResult => {
  if (!input.trim()) {
    return { ok: true, value: [], invalid: [] };
  }
  return normalizeItemProperties(input.split(","));
};

export const getItemPropertyLabel = (value: string, locale: string): string => {
  const slug = resolveItemPropertySlug(value);
  if (!slug) {
    return value.trim();
  }

  return locale === "pt"
    ? ITEM_PROPERTY_CONFIG[slug].labels.pt
    : ITEM_PROPERTY_CONFIG[slug].labels.en;
};

export const getItemPropertyLabels = (
  values: readonly string[] | null | undefined,
  locale: string,
) =>
  (values ?? [])
    .map((value) => getItemPropertyLabel(value, locale))
    .filter(Boolean);

File: server_py/app/services/item_properties.py
from __future__ import annotations

import re
import unicodedata
from enum import Enum

from app.models.base_item import BaseItemProperty

ITEM_PROPERTY_ALIASES: dict[str, tuple[str, ...]] = {
    "ammunition": ("ammunition", "ammo", "municao"),
    "finesse": ("finesse", "acuidade"),
    "heavy": ("heavy", "pesada", "pesado"),
    "light": ("light", "leve"),
    "loading": ("loading", "carregamento", "recarga"),
    "range": ("range", "alcance", "ranged", "distancia", "a distancia"),
    "reach": ("reach", "alcance estendido"),
    "special": ("special", "especial"),
    "thrown": ("thrown", "arremesso", "arremessavel"),
    "two_handed": ("two_handed", "two handed", "two-handed", "duas maos"),
    "versatile": ("versatile", "versatil"),
    "stealth_disadvantage": (
        "stealth_disadvantage",
        "stealth disadvantage",
        "disadvantage on stealth",
        "desvantagem em furtividade",
    ),
}

ITEM_PROPERTY_SLUGS = tuple(member.value for member in BaseItemProperty)
WEAPON_PROPERTY_SLUGS = tuple(
    member.value
    for member in BaseItemProperty
    if member is not BaseItemProperty.STEALTH_DISADVANTAGE
)


def _normalize_token(value: str) -> str:
    ascii_value = (
        unicodedata.normalize("NFKD", value or "")
        .encode("ascii", "ignore")
        .decode("ascii")
    )
    return re.sub(r"\s+", " ", ascii_value.lower().replace("_", " ").replace("-", " ")).strip()


ITEM_PROPERTY_ALIAS_TO_SLUG: dict[str, str] = {}
for slug, aliases in ITEM_PROPERTY_ALIASES.items():
    ITEM_PROPERTY_ALIAS_TO_SLUG[_normalize_token(slug)] = slug
    for alias in aliases:
        ITEM_PROPERTY_ALIAS_TO_SLUG[_normalize_token(alias)] = slug


def resolve_item_property_slug(value: str) -> str | None:
    normalized = _normalize_token(value)
    if not normalized:
        return None
    return ITEM_PROPERTY_ALIAS_TO_SLUG.get(normalized)


def normalize_item_properties(
    values: list[str] | tuple[str, ...] | None,
) -> tuple[list[str], list[str]]:
    normalized: list[str] = []
    invalid: list[str] = []
    seen: set[str] = set()

    for entry in values or []:
        source_value = entry.value if isinstance(entry, Enum) else entry
        raw_value = str(source_value or "").strip()
        if not raw_value:
            continue

        slug = resolve_item_property_slug(raw_value)
        if slug is None:
            if raw_value not in invalid:
                invalid.append(raw_value)
            continue

        if slug not in seen:
            seen.add(slug)
            normalized.append(slug)

    return normalized, invalid
